### PR TITLE
Adds macro implementations to commonly used traits.

### PR DESCRIPTION
**Pull request by vandomej**
If needed, all of these macros can be changed into ```#[derive()]``` macros. But in order to do this I believe we'd have to create another entire crate . I can look into it more if need be. Regardless, all the derive macros would act the same, they are just syntax sugar to make things shorter by a few lies so I don't think it's worth it yet.

### CODE REVIEW

File: src/game/actors/game_object.rs [hunk 1]
```diff
@@ -6,7 +6,7 @@ use game::map::tile::Tile;
 pub trait GameObject {
     fn get_position(&self) -> (i32, i32);
 
-    fn set_position(&mut self, position: (i32, i32));
+    fn set_position(&mut self, xy: (i32, i32));
```
Comment (devsebb) [inline src/game/actors/game_object.rs]: I would revert this or name it pos, I think it's more semantically valuable than the name `xy`

File: src/game/actors/game_object.rs [hunk 2]
```diff
@@ -23,8 +23,7 @@ pub trait GameObject {
         }
 
         for tile in tiles {
-            if tile.get_x() == proposed_x &&
-            tile.get_y() == proposed_y &&
+            if tile.get_xy() == position &&
```
Comment (devsebb) [inline src/game/actors/game_object.rs]: Same as above

File: src/game/actors/player.rs [hunk 3]
```diff
@@ -4,28 +4,28 @@ use tcod::colors;
...
-use game::actors::health;
-use game::actors::game_object;
+use game::actors::health::Health;
+use game::actors::game_object::GameObject;
 
 #[derive(Debug, Clone)]
 pub struct Player {
-    pub x: i32,
-    pub y: i32,
+    pub xy: (i32, i32),
```
Comment (devsebb) [inline src/game/actors/player.rs]: xy->position or something similar

File: src/game/actors/player.rs [hunk 4]
```diff
@@ -4,28 +4,28 @@ use tcod::colors;
...
+    arms: Vec<i32>,
     torso: i32,
-    legs: (i32, i32)
+    legs: Vec<i32>,
+    symbol: char
 }
 
 impl Player {
-    pub fn new(x: i32, y: i32) -> Player {
+    pub fn new(xy: (i32, i32)) -> Player {
```
Comment (devsebb) [inline src/game/actors/player.rs]: same as above

File: src/game/actors/player.rs [hunk 5]
```diff
@@ -51,17 +51,18 @@ impl Player {
             _ => {},
         }
 
-        let proposed_position = (self.x + proposed_x, self.y + proposed_y);
-        game_object::GameObject::move_object(self, tiles, proposed_position).unwrap()
+        let (x, y) = self.xy;
+        let proposed_position = (x + proposed_x, y + proposed_y);
```
Comment (devsebb) [inline src/game/actors/player.rs]:
```suggestion
        let proposed_position = (self.xy.0 + proposed_x, self.xy.1 + proposed_y);
```

File: src/game/map/mapgen.rs [hunk 6]
```diff
@@ -44,52 +42,50 @@ pub fn dummy_gen(map_width: i32, map_height: i32) -> (Map, Vec<Box<Tile>>) {
     return (map, tiles);
 }
 
-pub fn empty_gen(map_width: i32, map_height: i32) -> (Map, Vec<Box<Tile>>) {
+pub fn empty_gen((map_width, map_height): (i32, i32)) -> (Map, Vec<Box<Tile>>) {
     let mut tiles: Vec<Box<Tile>> = Vec::new();
     let mut map = Map::new(map_width, map_height);
 
     for h in 0..map_height {
```
Comment (devsebb) [inline src/game/map/mapgen.rs]: We should be consistent in what we call these. If it's map_w above, it should be map_w here, etc.

File: src/game/map/mapgen.rs [hunk 7]
```diff
@@ -44,52 +42,50 @@ pub fn dummy_gen(map_width: i32, map_height: i32) -> (Map, Vec<Box<Tile>>) {
...
-    draw_box(&mut map, &mut tiles, 0, 0, map_width - 1, map_height - 1);
+    draw_box(&mut map, &mut tiles, (0, 0), (map_width - 1, map_height - 1));
 
     return (map, tiles);
 }
 
-fn new_floor(map: &mut Map, tiles: &mut Vec<Box<Tile>>, x: i32, y: i32) {
-    let tile = Box::new(Floor::new(x, y));
-    map.set(x, y, tile.get_see_through(), tile.get_walkable());
+fn new_floor(map: &mut Map, tiles: &mut Vec<Box<Tile>>, xy: (i32, i32)) {
```
Comment (devsebb) [inline src/game/map/mapgen.rs]: xy->pos or something

File: src/game/map/mapgen.rs [hunk 8]
```diff
@@ -176,8 +172,8 @@ pub fn add_to_map(rooms: Vec<Room>, m: Map, t: Vec<Box<Tile>>) -> (Map, Vec<Box<
 }
 
 fn generate_connecting_hallway(r1: &Room, r2: &Room) -> (Room, Room) {
-    let center_r1: (i32, i32) = (r1.x + r1.w/2, r1.y + r1.h/2);
-    let center_r2: (i32, i32) = (r2.x + r2.w/2, r2.y + r2.h/2);
+    let center_r1: (i32, i32) = (r1.xy.0 + r1.wh.0/2, r1.xy.1 + r1.wh.1/2);
```
Comment (devsebb) [inline src/game/map/mapgen.rs]: same as above

File: src/game/map/mapgen.rs [hunk 9]
```diff
@@ -193,18 +189,14 @@ fn generate_connecting_hallway(r1: &Room, r2: &Room) -> (Room, Room) {
     };
 
     let hallway_horizontal = Room {
-        x: hallway_horiz_xy.0,
-        y: hallway_horiz_xy.1,
-        w: center_delta.0.abs() + 1,
-        h: 2,
+        xy: hallway_horiz_xy,
```
Comment (devsebb) [inline src/game/map/mapgen.rs]: same as above

File: src/game/map/tile.rs [hunk 10]
```diff
@@ -1,87 +1,87 @@
 use tcod::console::*;
 
 pub trait Tile {
-    fn new(x: i32, y: i32) -> Self where Self: Sized;
-    fn get_x(&self) -> i32;
-    fn get_y(&self) -> i32;
+    fn new(xy: (i32, i32)) -> Self where Self: Sized;
+    fn get_xy(&self) -> (i32, i32);
```
Comment (devsebb) [inline src/game/map/tile.rs]: get_xy-> get_pos?

File: src/game/map/tile.rs [hunk 11]
```diff
@@ -1,87 +1,87 @@
...
+                return self.symbol;
+            }
+        }
+    };
+}
 
 pub struct Wall {
-    pub x: i32,
-    pub y: i32,
+    pub xy: (i32, i32),
```
Comment (devsebb) [inline src/game/map/tile.rs]: same as above

File: src/game/map/tile.rs [hunk 12]
```diff
@@ -1,87 +1,87 @@
...
-    fn draw(&self, mut window: &Root) {
-        window.put_char(self.x, self.y, '#', BackgroundFlag::None);
-    }
-
-    fn clear(&self, mut window: &Root) {
-        window.put_char(self.x, self.y, ' ', BackgroundFlag::Set);
+impl Default for Wall {
+    fn default() -> Wall {
+        Wall {
+            xy: (0, 0),
```
Comment (devsebb) [inline src/game/map/tile.rs]: same as above

File: src/game/map/tile.rs [hunk 13]
```diff
@@ -1,87 +1,87 @@
...
+        }
     }
 }
 
+implement_tile!(Wall);
 
 pub struct Floor {
-    pub x: i32,
-    pub y: i32,
+    pub xy: (i32, i32),
```
Comment (devsebb) [inline src/game/map/tile.rs]: same as above

File: src/game/scene.rs [hunk 14]
```diff
@@ -59,7 +59,8 @@ impl Scene {
 
     fn draw_scene(&self, window: &Root) {
         for tile in &self.tiles {
-            if self.map.is_in_fov(tile.get_x(), tile.get_y()) {
+            let (x, y) = tile.get_xy();
+            if self.map.is_in_fov(x, y) {
```
Comment (devsebb) [inline src/game/scene.rs]:
```suggestion
            if self.map.is_in_fov(xy.0, xy.1) {
```
Although I'd also recommend changing xy to pos or position.

**Comment by vandomej** [comment]
> The macros are good! Now that I'm looking at it, though, it seems like the pairing together of the x and y coordinates/width and height complicates calling the functions a bit. For instance, there's a lot of
> 
> ```
> foo((x, y));
> ```
> 
> ...now, which strikes me as a bit ugly. What do you think -- is it worth keeping that part of the pull request because it pairs related values together, or is it too difficult to call functions now?
> 
> If we do keep it, I'd recommend renaming the parameters. You have a lot of `xy`, which I think would be better expressed as `pos` or `position`.
> 
> In any case, once we come to a decision about the above you can just merge this branch, since there doesn't look like there's going to be much logic changing.

It is true that it complicates a lot of function signatures. I think that keeping them as Tuples or keeping them separate is fine as long as it's consistent throughout the entire application. As far as keeping them separate, that would make sense since a lot of tcod's functions have them separate as parameters. But at the same time there are functions which return them as Tuples so it's a bit messy regardless of what we choose. Also another thing to keep in mind about Tuple's is that they are useless until you obtain their values, which often leads to:

```
//Destruct the tuple
fun ((x, y): (i32, i32)) {
//use x and y
//..

//Construct tuple again
fun2((x, y))
}
```

This doesn't seem like a win-win scenario, with Tuple's we'll get some messy and unsatisfying syntax, and without Tuple's we will have a slightly unpleasing assemblage of data. But i'm more inclined to keep them separate at this point and might add a commit doing the opposite.

**Comment by devsebb** [comment]
> > The macros are good! Now that I'm looking at it, though, it seems like the pairing together of the x and y coordinates/width and height complicates calling the functions a bit. For instance, there's a lot of
> > ```
> > foo((x, y));
> > ```
> > 
> > 
> > ...now, which strikes me as a bit ugly. What do you think -- is it worth keeping that part of the pull request because it pairs related values together, or is it too difficult to call functions now?
> > If we do keep it, I'd recommend renaming the parameters. You have a lot of `xy`, which I think would be better expressed as `pos` or `position`.
> > In any case, once we come to a decision about the above you can just merge this branch, since there doesn't look like there's going to be much logic changing.
> 
> It is true that it complicates a lot of function signatures. I think that keeping them as Tuples or keeping them separate is fine as long as it's consistent throughout the entire application. As far as keeping them separate, that would make sense since a lot of tcod's functions have them separate as parameters. But at the same time there are functions which return them as Tuples so it's a bit messy regardless of what we choose. Also another thing to keep in mind about Tuple's is that they are useless until you obtain their values, which often leads to:
> 
> ```
> //Destruct the tuple
> fun ((x, y): (i32, i32)) {
> //use x and y
> //..
> 
> //Construct tuple again
> fun2((x, y))
> }
> ```
> 
> This doesn't seem like a win-win scenario, with Tuple's we'll get some messy and unsatisfying syntax, and without Tuple's we will have a slightly unpleasing assemblage of data. But i'm more inclined to keep them separate at this point and might add a commit doing the opposite.

One thing that I thought of doing was using structs, so that you just end up doing something like this:

```
let p = Point {
    x: 2,
    y: 2
}

foo(*p);
```

or even

```
foo(*Point{x: 2, y:2});
```

...but I feel like that )especially the second option) might just complicate things further, unfortunately. I think it'd be a good idea to just keep them separate, for now, and deal with destructuring tcod's tuples manually.

**Pull request merged**